File: C++/Mediator/WidgetManager/widget.cc
#include <iostream>
#include "widget.h"
#include "widget_manager.h"

Widget::Widget(WidgetManager* widget_manager) : _widget_manager(widget_manager) {}

void Widget::changed() {
    _widget_manager->widget_changed(this);
}


TextBox::TextBox(WidgetManager* widget_manager, const std::string& text) : Widget(widget_manager), _text(text) {}

void TextBox::set_text(const std::string &text) {
    std::cout << "Setting text to \"" << text << "\" ...\n";
    _text = text;
    changed();
}

const std::string& TextBox::get_text() const {
    return _text;
}


DisplayWindow::DisplayWindow(WidgetManager* widget_manager) : Widget(widget_manager) {}

void DisplayWindow::display(const std::string& text) const {
    std::string right_pad(76 - text.size(), ' ');
    right_pad += "||";
    std::cout << "Displaying text...\n";
    std::cout << std::string(80, '=') << '\n';
    std::cout << "||" << std::string(76, ' ') << "||\n";
    std::cout << "||" << text << right_pad << '\n';
    std::cout << "||" << std::string(76, ' ') << "||\n";
    std::cout << std::string(80, '=') << '\n';
}


Button::Button(WidgetManager* widget_manager, const std::string& name) : Widget(widget_manager), _name(name) {}

void Button::clicked() {
    std::cout << "Button " << _name << " clicked...\n";
    changed();
}

File: Java/Mediator/WidgetManager/src/com/ericgtkb/Widget.java
package com.ericgtkb;

import java.util.Arrays;

public abstract class Widget {
    protected WidgetManager manager;

    protected Widget(WidgetManager manager) {
        this.manager = manager;
    }

    void changed() {
        manager.widgetChanged(this);
    }
}


class TextBox extends Widget {
    private String text;

    public TextBox(WidgetManager manager, String text) {
        super(manager);
        this.text = text;
    }

    public void setText(String text) {
        System.out.println("Setting text to \"" + text + "\"...");
        this.text = text;
        // Call manager when new text is set
        changed();
    }

    public String getText() {
        return text;
    }
}

class DisplayWindow extends Widget {
    public DisplayWindow(WidgetManager manager) {
        super(manager);
    }

    void display(String text) {
        char[] spaces = new char[76 - text.length()];
        char s = ' ';
        Arrays.fill(spaces, s);
        String rightPad = new String(spaces) + "||";
        System.out.println("Displaying text...");
        System.out.println("================================================================================");
        System.out.println("||                                                                            ||");
        System.out.println("||" + text + rightPad);
        System.out.println("||                                                                            ||");
        System.out.println("================================================================================");
    }
}

class Button extends Widget {
    private String name;

    public Button(WidgetManager manager, String name) {
        super(manager);
        this.name = name;
    }

    public void clicked() {
        System.out.println("Button " + name + " clicked...");
        // Call manager to handle event
        changed();
    }
}


